File: cpp/day1.cpp
// Copyright 2023 Wojciech Sek

#include <cassert>
#include "utils.hpp"

int to_num(char c) {
    switch (c) {
        case '(': return 1;
        case ')': return -1;
        default: return 0;
    }
}

int task1(const std::string &s) {
    int counter = 0;

    for (char c : s) {
        counter += to_num(c);
    }

    return counter;
}

int task2(const std::string &s) {
    int counter = 0;
    int length = s.length();

    for (int i = 0; i < length; i++) {
        switch (counter) {
            case -1: return i;
            default: counter += to_num(s[i]);
        }
    }

    return length;
}

int main() {
    std::string s = file_to_string("../input/day1.txt");

    assert(task1(s) == 138);
    assert(task2(s) == 1771);

    return 0;
}

File: ocaml/day1.ml
let to_num c =
  match c with 
  | '(' -> 1
  | ')' -> -1
  | _   -> failwith "Unknown character";;

let task1 chars =
  chars
  |> List.map to_num 
  |> List.fold_left (+) 0;;

let task2 chars =
    let rec aux floor pos list =
      match (floor, list) with 
      | (-1, _)
      | (_, []) -> pos
      | (_, c::cs) -> aux (floor + to_num c) (pos + 1) cs
  in 
  chars |> aux 0 0;;

  let chars = Utils.file_to_chars ("../input/day1.txt") in
  let result1 = task1 chars in
  let result2 = task2 chars in
    Printf.printf "Task1: %d\n" result1;
    Printf.printf "Task2: %d\n" result2;
    assert (result1 = 138);
    assert (result2 = 1771);